## 6*6 Case 

In [2]:
import GridWorldV2
import MDP
import LPSOS1   
import numpy as np

mdp, V_def, policy = GridWorldV2.createGridWorldBarrier_new2()#new3 to new2 to change case
st_len = len(mdp.statespace)
D, E, F = LPSOS1.generate_matrix(mdp)
k = 4
init = np.zeros(st_len)
init[12] = 1

LPSOS1.LP(mdp, k, init)

Set parameter Username
Academic license - for non-commercial use only - expires 2025-02-20
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Start optimization
Set parameter NodeLimit to value 1073741824
Set parameter SolutionLimit to value 1073741824
Set parameter IntFeasTol to value 1e-06
Set parameter Method to value 3
Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.3.0 23D60)

CPU model: Apple M2
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 503 rows, 360 columns and 1468 nonzeros
Model fingerprint: 0xa3f74af2
Model has 144 SOS constraints
Variable types: 360 continuous, 0 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e-01, 1e+00]
  Objective range  [1e+00, 1e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 4e+00]
Presolve removed 338 rows and 58 columns
Presolve time: 0.00s
Presolved: 165 rows, 302 columns, 911 nonzeros
Presolved model has 13

In [3]:
import chebyshev_center_v2
import pickle

with open('x_res', 'rb') as fp:
    x_res = pickle.load(fp)

with open('y_res', 'rb') as fp:
    y_res = pickle.load(fp)

center, gamma, m_star = chebyshev_center_v2.chebyshev(y_res,k,mdp,init)
print(center, gamma)


Set parameter Username
Academic license - for non-commercial use only - expires 2025-02-20
[2.12618751 1.87381249] 0.08738182826222071


In [ ]:
import reward_allocation

c = 0.086
opt = 0.4326
reward_allocation.LP(mdp, k, init, c, opt)

Set parameter Username
Academic license - for non-commercial use only - expires 2025-02-20
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Start optimization
Set parameter NodeLimit to value 1073741824
Set parameter SolutionLimit to value 1073741824
Set parameter IntFeasTol to value 1e-06
Set parameter Method to value 3
Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.3.0 23D60)

CPU model: Apple M2
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 1368 rows, 900 columns and 3971 nonzeros
Model fingerprint: 0xd87f649c
Model has 576 SOS constraints
Variable types: 900 continuous, 0 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e-01, 1e+00]
  Objective range  [0e+00, 0e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [9e-02, 4e+00]
Presolve removed 815 rows and 106 columns
Presolve time: 0.00s
Presolved: 553 rows, 794 columns, 2470 nonzeros
Presolved model has

0.174 infeasible 1m35s
0.1653 infeasible 6m14s
0.1566 infeasible 2m16s
0.088 infeasible 3m54s
0.087 feasible *** 0.2s
0.086 feasible 0.3s


In [ ]:
import pickle
with open('x_res_2', 'rb') as fp:
    x_res_2 = pickle.load(fp)

with open('y_res_2', 'rb') as fp:
    y_res_2 = pickle.load(fp)

decoy_value_2 = reward_allocation.extract_decoy_value(x_res_2, mdp)
print(decoy_value_2)

[2.12164046 1.86930254]


### Robustness check

In [ ]:
import unique_BR
import pickle
import reward_allocation
tau = 1e-10

with open('x_res', 'rb') as fp:
    x_res = pickle.load(fp)

with open('y_res', 'rb') as fp:
    y_res = pickle.load(fp)

with open('x_res_2', 'rb') as fp:
    x_res_2 = pickle.load(fp)

with open('y_res_2', 'rb') as fp:
    y_res_2 = pickle.load(fp)

decoy_value_2 = reward_allocation.extract_decoy_value(x_res_2, mdp)
decoy_value = reward_allocation.extract_decoy_value(x_res, mdp)
pess_1, opt_1 = unique_BR.is_unique(reward_allocation.from_x_to_R2(decoy_value, mdp), tau, mdp, y_res, init)
pess_2, opt_2 = unique_BR.is_unique(reward_allocation.from_x_to_R2(decoy_value_2, mdp), tau, mdp, y_res_2, init)
print(pess_1, opt_1)
print(pess_2, opt_2)

0.39816462607991665 0.43260375609784024
0.4326037550532157 0.43260375609784035


tau = 1e-1
0.0 0.43260375609784024
0.006174760571561906 0.43260375609784024
tau = 1e-2 
0.07142571209688497 0.43260375609784024
0.37855218655363443 0.43260375609784024
tau = 1e-3
0.10743517253237922 0.43260375609784024
0.42478147872045613 0.43260375609784024
tau = 1e-4
0.19463776243204514 0.43260375609784024
0.43156218420366876 0.43260375609784024
tau = 1e-5
0.37781212289022614 0.43260375609784024
0.43249929353524874 0.43260375609784024
tau = 1e-6
0.39612955893497975 0.43260375609784024
0.4325933098415797 0.43260375609784024
tau = 1e-7
0.39796130254100776 0.43260375609784024
0.432602711472216 0.43260375609784035
tau = 1e-8
0.39814447690037713 0.43260375609784024
0.43260365163527825 0.43260375609784035
tau = 1e-9
0.3981627943361226 0.43260375609784024
0.4326037456515854 0.43260375609784035
tau = 1e-10
0.39816462607991665 0.43260375609784024
0.4326037550532157 0.43260375609784035

In [46]:
import gradient_new
import cyipopt_optimize

tau = 1e-5
x_grad_np, y_np, def_value = gradient_new.ent_gradient(mdp, tau, decoy_value_2,init, y_test=np.array(y_res_2), debug=True)
x_grad_np, y_np, def_value = gradient_new.ent_gradient(mdp, tau, decoy_value,init, y_test=np.array(y_res), debug=True)
cyipopt_optimize.test_cyipopt_optimize(False, False, tau, init, mdp, x_res_2)


Feasibility of y_test (SSE): 2.220446049250313e-16
Attacker value under y_test: 0.9130476140119977
Defender value under y_test: 0.43260375609784035
Attacker opt value (boundedly rational response): 0.9130471594142585
Defender opt value (boundedly rational response): 0.43260329519716456
Feasibility of y_test (SSE): 2.220446049250313e-16
Attacker value under y_test: 0.8406307642489206
Defender value under y_test: 0.43260375609784024
Attacker opt value (boundedly rational response): 0.8406305291433362
Defender opt value (boundedly rational response): 0.41372193208794894
[2.12164046 1.86930254]

******************************************************************************
This program contains Ipopt, a library for large-scale nonlinear optimization.
 Ipopt is released as open source code under the Eclipse Public License (EPL).
         For more information visit https://github.com/coin-or/Ipopt
******************************************************************************

This is Ipopt v

tau = 1e-1
0.0012938646617501207
0.000863
0.001311174155800507 
tau = 1e-2
0.42876004484833513
0.3346524146117557
0.4292259984592307
tau = 1e-3
0.43260335988108656
0.323738746534165
0.432603376384169
tau = 1e-4
0.43260556825943763
0.38519708875819014
0.43260513806877665
tau = 1e-5
0.43260513806877665
0.43260513806877665
0.41372193208794894

## 10*10 Case

In [92]:
import GridWorldV2
mdp, V_def, policy = GridWorldV2.createGridWorldBarrier_new3()#new3 to new2 to change case
st_len = len(mdp.statespace)
init = np.zeros(st_len)
init[30] = 1
LPSOS1.LP(mdp, k, init)

0.48338892293408764 0.5153067021020193
Set parameter Username
Academic license - for non-commercial use only - expires 2025-02-20
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Start optimization
Set parameter NodeLimit to value 1073741824
Set parameter SolutionLimit to value 1073741824
Set parameter IntFeasTol to value 1e-06
Set parameter Method to value 3
Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.3.0 23D60)

CPU model: Apple M2
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 1398 rows, 1000 columns and 4379 nonzeros
Model fingerprint: 0x5e781c7a
Model has 400 SOS constraints
Variable types: 1000 continuous, 0 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e-01, 1e+00]
  Objective range  [1e+00, 1e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 4e+00]
Presolve removed 919 rows and 127 columns
Presolve time: 0.00s
Presolved: 479 rows, 873 c

In [93]:
import chebyshev_center_v2
import pickle

with open('x_res', 'rb') as fp:
    x_res = pickle.load(fp)

with open('y_res', 'rb') as fp:
    y_res = pickle.load(fp)

center, gamma, m_star = chebyshev_center_v2.chebyshev(y_res,k,mdp,init)
print(x_res)
print(center, gamma)

[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.4287144394630527, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.466390388300294, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
[2.00500502 0.         1.99499498] 0.06126009920621299


In [ ]:
import reward_allocation

c = 0.061
opt = 0.49496
reward_allocation.LP(mdp, k, init, c, opt)

Set parameter Username
Academic license - for non-commercial use only - expires 2025-02-20
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Start optimization
Set parameter NodeLimit to value 1073741824
Set parameter SolutionLimit to value 1073741824
Set parameter IntFeasTol to value 0.0001
Set parameter FeasibilityTol to value 0.0001
Set parameter Method to value 3
Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.3.0 23D60)

CPU model: Apple M2
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 5399 rows, 3500 columns and 16749 nonzeros
Model fingerprint: 0x73374dd2
Model has 2400 SOS constraints
Variable types: 3500 continuous, 0 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e-01, 1e+00]
  Objective range  [0e+00, 0e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [6e-02, 4e+00]
Presolve removed 3023 rows and 230 columns
Presolve time: 0.01s
Presolved: 2376 

In [ ]:
import reward_allocation

c = 0.194
opt = 0.49496
reward_allocation.LP(mdp, k, init, c, opt)

Set parameter Username
Academic license - for non-commercial use only - expires 2025-02-20
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Start optimization
Set parameter NodeLimit to value 1073741824
Set parameter SolutionLimit to value 1073741824
Set parameter IntFeasTol to value 0.0001
Set parameter FeasibilityTol to value 0.0001
Set parameter Method to value 3
Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.3.0 23D60)

CPU model: Apple M2
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 5399 rows, 3500 columns and 16749 nonzeros
Model fingerprint: 0x1079d632
Model has 2400 SOS constraints
Variable types: 3500 continuous, 0 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e-01, 1e+00]
  Objective range  [0e+00, 0e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e-01, 4e+00]
Presolve removed 3023 rows and 230 columns
Presolve time: 0.01s
Presolved: 2376 

0.07 optimal [1.997828 0.       2.002172]             2.2s
0.1 optimal [1.42736, 0, 1.56638]                     2.1s
1 infeasible                                          1.9s
0.5 infeasible                                        2.4s
0.3 infeasible                                        1m2s
0.2 infeasible                                        1min25s                     
0.195 infeasible                                      14.2s
0.194 feasible [1.83588357 0.         2.16411643] *** 2.6s
0.19 feasible                                         2.5s
0.18 feasible                                         2.2s
0.17 feasible                                         4.6s
0.15 feasible                                         2.3s
0.13 feasible [1.7987134  0.         1.97788684]      2.1s
0.14 feasible [1.90691096 0.         2.09308904]      2.4s

In [ ]:
import pickle

with open('x_res_2', 'rb') as fp:
    x_res_2 = pickle.load(fp)

with open('y_res_2', 'rb') as fp:
    y_res_2 = pickle.load(fp)

decoy_value = reward_allocation.extract_decoy_value(x_res_2, mdp)
print(decoy_value)
print(y_res_2)

[1.83588357 0.         2.16411643]
[0.0, 0.0, 0.0, 0.0, 1.4302269728331332e-06, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.3487684822950075e-05, 0.0, 0.0, 0.0, 2.0149651604296663e-05, 1.914216902408183e-06, 0.0, 0.0, 0.0, 1.0849594071561081e-05, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 2.196987250884596e-05, 2.0871378883403663e-06, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00011038416232830223, 0.0, 0.0, 0.0, 0.00019861391101175165, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00022729936212529232, 8.138678680444645e-05, 0.0, 0.0, 0.0, 0.00014446506331010041, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.00016205733587916577, 0.0, 0.0, 0.0, 1.5719638958410114e-05, 0.0, 0.0, 0.0, 3.412547893572287e-06, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0010529402004735824, 0.0, 0.0, 0.0, 0.0008191028739060392, 0.0, 0.0, 0.0, 0.0008037135763827487, 0.0, 0.0, 0.0, 0.0006185540627973114, 0.0, 0.0, 0.0, 0.0005317742665629197, 0.0, 0.0, 0.0, 0.0005344265480256893, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 

### Robustness Check

In [ ]:
import unique_BR
import pickle
import reward_allocation
tau = 1e-10

with open('x_res', 'rb') as fp:
    x_res = pickle.load(fp)

with open('y_res', 'rb') as fp:
    y_res = pickle.load(fp)

with open('x_res_2', 'rb') as fp:
    x_res_2 = pickle.load(fp)

with open('y_res_2', 'rb') as fp:
    y_res_2 = pickle.load(fp)

decoy_value_2 = reward_allocation.extract_decoy_value(x_res_2, mdp)
decoy_value = reward_allocation.extract_decoy_value(x_res, mdp)
pess_1, opt_1 = unique_BR.is_unique(reward_allocation.from_x_to_R2(decoy_value, mdp), tau, mdp, y_res, init)
pess_2, opt_2 = unique_BR.is_unique(reward_allocation.from_x_to_R2(decoy_value_2, mdp), tau, mdp, y_res_2, init)
print(pess_1, opt_1)
print(pess_2, opt_2)

0.4944946851621991 0.49496017497668976
0.49491421016400583 0.49493197698567354


tau = 1e-1
0.298824907398692 0.49496017497668976
0.4096599278795998 0.49493197698567354
tau = 1e-2
0.4702750797051762 0.49496017497668976
0.4850997971464766 0.49493197698567354
tau = 1e-3
0.49118003685981443 0.49496017497668976
0.49365768138351224 0.49493197698567354
tau = 1e-4
0.49394151993370505 0.49496017497668976
0.4947645078186484 0.49493197698567354
tau = 1e-5
0.49441163821212997 0.49496017497668976
0.4948980552188916 0.49493197698567354
tau = 1e-6
0.49446672870051883 0.49496017497668976
0.4949125436800391 0.49493197698567354
tau = 1e-7
0.494491892032983 0.49496017497668976
0.49491404366560576 0.49493197698567354
tau = 1e-8
0.49449440836609493 0.49496017497668976
0.4949141936641639 0.49493197698567354
tau = 1e-9
0.49449465999758224 0.49496017497668976
0.49491420866402 0.49493197698567354
tau = 1e-10
0.4944946851621991 0.49496017497668976
0.49491421016400583 0.49493197698567354

In [14]:
import gradient_new
import cyipopt_optimize

tau = 1e-1
x_grad_np, y_np, def_value = gradient_new.ent_gradient(mdp, tau, decoy_value_2,init, y_test=np.array(y_res_2), debug=True)
x_grad_np, y_np, def_value = gradient_new.ent_gradient(mdp, tau, decoy_value,init, y_test=np.array(y_res), debug=True)
cyipopt_optimize.test_cyipopt_optimize(False, False, tau, init, mdp, x_res_2)

Feasibility of y_test (SSE): 7.765906840327904e-07
Attacker value under y_test: 1.0709396249573835
Defender value under y_test: 0.49493197698567354
Attacker opt value (boundedly rational response): 0.002663841249823807
Defender opt value (boundedly rational response): 0.0012086245707876838
Feasibility of y_test (SSE): 1.6981731396081834e-08
Attacker value under y_test: 0.7257764680190252
Defender value under y_test: 0.49496017497668976
Attacker opt value (boundedly rational response): 0.00038242746050301546
Defender opt value (boundedly rational response): 0.00017782435918724747
[1.83588357 0.         2.16411643]
This is Ipopt version 3.14.14, running with linear solver MUMPS 5.6.2.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        3
Number of nonzeros in Lagrangian Hessian.............:        0

Total number of variables............................:        3
                     variables with only lower bou

/Users/shuowu/miniconda3/envs/robust_allocation/lib/python3.10/site-packages/diffcp/cone_program.py:295: UserWarning: Solved/Inaccurate.
  warnings.warn("Solved/Inaccurate.")


 155 -2.6297683e-02 0.00e+00 4.32e-02 -11.0 3.71e-02    -  1.00e+00 1.00e+00f  1
 156 -2.9881053e-02 0.00e+00 4.85e-02 -11.0 4.12e-02    -  1.00e+00 1.00e+00f  1
 157 -3.4380214e-02 0.00e+00 5.47e-02 -11.0 4.60e-02    -  1.00e+00 1.00e+00f  1
 158 -4.0102116e-02 0.00e+00 6.21e-02 -11.0 5.18e-02    -  1.00e+00 1.00e+00f  1
 159 -4.7449647e-02 0.00e+00 7.07e-02 -11.0 5.87e-02    -  1.00e+00 1.00e+00f  1
iter    objective    inf_pr   inf_du lg(mu)  ||d||  lg(rg) alpha_du alpha_pr  ls
 160 -5.6956706e-02 0.00e+00 8.05e-02 -11.0 6.67e-02    -  1.00e+00 1.00e+00f  1
 161 -6.9255981e-02 0.00e+00 9.12e-02 -11.0 7.59e-02    -  1.00e+00 1.00e+00f  1
 162 -8.5033943e-02 0.00e+00 1.02e-01 -11.0 8.61e-02    -  1.00e+00 1.00e+00f  1
 163 -1.0484675e-01 0.00e+00 1.07e-01 -11.0 9.69e-02    -  1.00e+00 1.00e+00f  1
 164 -1.2790206e-01 0.00e+00 1.20e-01 -11.0 1.03e-01    -  1.00e+00 1.00e+00f  1
 165 -1.5499138e-01 0.00e+00 1.22e-01 -11.0 1.15e-01    -  1.00e+00 1.00e+00f  1
 166 -1.8376097e-01 0.00e+00

## Attack Graph Case

In [15]:
policy, V_att, V_def, st_visit, mdp = MDP.test_att()
st_len = len(mdp.statespace)
init = np.zeros(st_len)
init[0] = 1
LPSOS1.LP(mdp, k, init)

Itcount: 1
Itcount: 2
Itcount: 3
Itcount: 4
Itcount: 5
Itcount: 6
Itcount: 7
Itcount: 8
Set parameter Username
Academic license - for non-commercial use only - expires 2025-02-20
Start optimization
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Set parameter NodeLimit to value 1073741824
Set parameter SolutionLimit to value 1073741824
Set parameter IntFeasTol to value 1e-06
Set parameter Method to value 3
Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.3.0 23D60)

CPU model: Apple M2
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 209 rows, 150 columns and 572 nonzeros
Model fingerprint: 0xd0ae7013
Model has 60 SOS constraints
Variable types: 150 continuous, 0 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e-01, 1e+00]
  Objective range  [1e+00, 1e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 4e+00]
Presolve removed 148 rows and 36 columns
Pres

In [16]:
import chebyshev_center_v2
import pickle

with open('x_res', 'rb') as fp:
    x_res = pickle.load(fp)

with open('y_res', 'rb') as fp:
    y_res = pickle.load(fp)

center, gamma, m_star = chebyshev_center_v2.chebyshev(y_res,k,mdp,init)
print(center, gamma)

[4. 0.] 2.781954887218044


In [ ]:
import reward_allocation

c = 1.34
opt = 0.6545664
reward_allocation.LP(mdp, k, init, c, opt)

Set parameter Username
Academic license - for non-commercial use only - expires 2025-02-20
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Start optimization
Set parameter NodeLimit to value 1073741824
Set parameter SolutionLimit to value 1073741824
Set parameter IntFeasTol to value 0.0001
Set parameter FeasibilityTol to value 0.0001
Set parameter Method to value 3
Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.3.0 23D60)

CPU model: Apple M2
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 570 rows, 375 columns and 1563 nonzeros
Model fingerprint: 0xafdecffd
Model has 240 SOS constraints
Variable types: 375 continuous, 0 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e-01, 1e+00]
  Objective range  [0e+00, 0e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e-01, 4e+00]
Presolve removed 365 rows and 79 columns
Presolve time: 0.00s
Presolved: 205 rows, 29

In [ ]:
import reward_allocation

c = 1.33
opt = 0.6545664
reward_allocation.LP(mdp, k, init, c, opt)

Set parameter Username
Academic license - for non-commercial use only - expires 2025-02-20
Warning for adding constraints: zero or small (< 1e-13) coefficients, ignored
Start optimization
Set parameter NodeLimit to value 1073741824
Set parameter SolutionLimit to value 1073741824
Set parameter IntFeasTol to value 0.0001
Set parameter FeasibilityTol to value 0.0001
Set parameter Method to value 3
Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.3.0 23D60)

CPU model: Apple M2
Thread count: 8 physical cores, 8 logical processors, using up to 8 threads

Optimize a model with 570 rows, 375 columns and 1563 nonzeros
Model fingerprint: 0xb69a202a
Model has 240 SOS constraints
Variable types: 375 continuous, 0 integer (0 binary)
Coefficient statistics:
  Matrix range     [1e-01, 1e+00]
  Objective range  [0e+00, 0e+00]
  Bounds range     [0e+00, 0e+00]
  RHS range        [7e-01, 4e+00]
Presolve removed 365 rows and 79 columns
Presolve time: 0.00s
Presolved: 205 rows, 29

In [ ]:
import pickle

with open('x_res_2', 'rb') as fp:
    x_res_2 = pickle.load(fp)

with open('y_res_2', 'rb') as fp:
    y_res_2 = pickle.load(fp)

decoy_value_2 = reward_allocation.extract_decoy_value(x_res_2, mdp)
decoy_value =  reward_allocation.extract_decoy_value(x_res, mdp)
print(decoy_value_2)
print(decoy_value)

[2.66 1.33]
[1.21804511 0.        ]


In [ ]:
import unique_BR
import pickle
import reward_allocation
tau = 1e-6

with open('x_res', 'rb') as fp:
    x_res = pickle.load(fp)

with open('y_res', 'rb') as fp:
    y_res = pickle.load(fp)

with open('x_res_2', 'rb') as fp:
    x_res_2 = pickle.load(fp)

with open('y_res_2', 'rb') as fp:
    y_res_2 = pickle.load(fp)

decoy_value_2 = reward_allocation.extract_decoy_value(x_res_2, mdp)
decoy_value = reward_allocation.extract_decoy_value(x_res, mdp)
pess_1, opt_1 = unique_BR.is_unique(reward_allocation.from_x_to_R2(decoy_value, mdp), tau, mdp, y_res, init)
pess_2, opt_2 = unique_BR.is_unique(reward_allocation.from_x_to_R2(decoy_value_2, mdp), tau, mdp, y_res_2, init)
print(pess_1, opt_1)
print(pess_2, opt_2)

/Users/shuowu/miniconda3/envs/robust_allocation/lib/python3.10/site-packages/cvxpy/problems/problem.py:1409: UserWarning: 
    The problem is either infeasible or unbounded, but the solver
    cannot tell which. Disable any solver-specific presolve methods
    and re-solve to determine the precise problem status.

    For GUROBI and CPLEX you can automatically perform this re-solve
    with the keyword argument prob.solve(reoptimize=True, ...).
    
  warnings.warn(INF_OR_UNB_MESSAGE)


ValueError: matmul: Input operand 1 does not have enough dimensions (has 0, gufunc core with signature (n?,k),(k,m?)->(n?,m?) requires 1)

In [27]:
import gradient_new
import cyipopt_optimize

tau = 1e-5
x_grad_np, y_np, def_value = gradient_new.ent_gradient(mdp, tau, decoy_value_2,init, y_test=np.array(y_res_2), debug=True)
x_grad_np, y_np, def_value = gradient_new.ent_gradient(mdp, tau, decoy_value,init, y_test=np.array(y_res), debug=True)
cyipopt_optimize.test_cyipopt_optimize(False, False, tau, init, mdp, x_res_2)

Feasibility of y_test (SSE): 5.654642599994819e-07
Attacker value under y_test: 1.70454416749228
Defender value under y_test: 0.6545664
Attacker opt value (boundedly rational response): 1.7045437252831728
Defender opt value (boundedly rational response): 0.6545665785498868
Feasibility of y_test (SSE): 1.1102230246251565e-16
Attacker value under y_test: 0.7698492399420311
Defender value under y_test: 0.6545664886257698
Attacker opt value (boundedly rational response): 0.7698504282547531
Defender opt value (boundedly rational response): 0.5006620164197537
[2.66 1.33]
This is Ipopt version 3.14.14, running with linear solver MUMPS 5.6.2.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        2
Number of nonzeros in Lagrangian Hessian.............:        0

Total number of variables............................:        2
                     variables with only lower bounds:        2
                variables with lowe